drafts/Asynchronous-programming-made-resilient-with-MPFT.adoc

```diff
@@ -0,0 +1,237 @@
+
+== Deep dive: Asynchronous flow of execution
+This last section isn't critical for knowing how to use `@Asynchronous`, but it's useful to know when working with MicroProfile Fault Tolerance. It covers how the flow of execution changes when using the `@Asynchronous` annotation compared to when it's not used. 
+
+Let's first look at how the Fault Tolerance annotations (`@Retry`, `@Timeout`, `@CircuitBreaker`, `@Bulkhead`,  and `@Fallback`) work together *without* the presence of `@Asynchronous` :
+
+image::/img/blog/FT-synchronous-execution-flow.png[Fault Tolerance synchronous execution flow]
+
+When the method is annotated with `@Asynchronous` a few things change. The differences from synchronous execution are highlighted in dark green, and discussed below:
+
+[caption="Accurate for mpFaultTolerance-2.0"]
+image::/img/blog/FT-asynchronous-execution-flow.png[Fault Tolerance asynchronous execution flow]
+
+The first difference compared to the synchronous flow of execution is that a `CompletionStage` or `Future` is returned before the method runs. When the method has actually returned, the result from the method is then propagated into the `CompletionStage` or `Future` so that the caller can get it. 
+
+The next difference comes in the Bulkhead. As well as either accepting or rejecting the execution, the Bulkhead can also queue it to be run later. If the method is accepted by the Bulkhead, it is then scheduled to be run on another thread, rather than run immediately. 
+
+Another difference occurs with Timeout. When used with @Asynchronous, if the Timeout expires, the method is interrupted, and the execution skips forwards to the point highlighted in the diagram (From the "Timeout Expires" block). The result is then processed as if the method had finished with a TimeoutException.
+
+The last difference is that if there's a Fallback, it also runs asynchronously, so it's scheduled to run on another thread as well.
+
+[#CS-preference]
+=== Why is `CompletionStage` prefered?
+If a method returns a `CompletionStage` rather than a `Future`, then there's one more useful difference compared to the synchronous flow of execution. If the method returns a value rather than throwing an exception, the execution waits until that `CompletionStage` completes before doing the rest of the Fault Tolerance logic. There are a few benefits to point out here: 
+
+* The execution reserves a space on the Bulkhead and doesn't release it until after the returned `CompletionStage` completes. So, as far as the Bulkhead is concerned, it's still "executing" until the `CompletionStage` is complete. 
+* The timeout starts before we check if there’s space on the Bulkhead, at which point it might be queued. If it is queued, the time it spends queuing still counts towards the timeout. 
+* Similarly, the timeout isn't stopped until after the `CompletionStage` completes. 
```

Comment:
* The middle point is not unique to `CompletionStage`, the timeout is always started before an execution is queued.
* There's a missing point: if the method returns a `CompletionStage` that completes with an exception, Retry, CircuitBreaker and Fallback will act as if that exception was thrown from the method.